### PR TITLE
Distribution guide

### PR DESCRIPTION
**Pull request by qlzh727**
Add the API guide for the new unified distribution API.

This is the first draft, and let's iterate on this one.

### CODE REVIEW

File: guides/unified_distribution.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+The Keras distribution API is a new API interface designed to facilitate 
+distributed deep learning across a variety of backends like JAX, Tensorflow and
+Pytorch. This innovative API introduces a suite of tools enabling data and model
+parallelism, allowing for efficient scaling of deep learning models on multiple
+accelerators and hosts. Whether leveraging the power of GPUs or TPUs, the API 
+provides a streamlined approach to initializing distributed environments, 
+defining device meshes, and orchestrating the layout of tensors across 
+computational resources. Through classes like `DataParallel` and 
+`ModelParallel`, it abstracts the complexity involved in parallel computation, 
+making it accessible for developers to accelerate their machine learning
```
Comment (pcoet) [inline guides/unified_distribution.py:29]: "making it accessible for" -> "enabling"

File: guides/unified_distribution.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+"""
+ModelParallel and LayoutMap
+
+ModelParallel will be mostly useful when model weights are too large to fit
+into a single accelerator. This setting allows you to spit you model weights or
+activation tensors across all the devices on the DeviceMesh, and enable the 
+horizontal scaling for the large models.
+
+Unlike the `DataParallel` model that all weights are fully replicated,
```
Comment (pcoet) [inline guides/unified_distribution.py:167]: "model that" -> "model where"

**Review by pcoet — COMMENTED** [review]
Helpful doc! I made a bunch of nitty comments, but ... they're nits. :-)

File: guides/unified_distribution.py (hunk 1 above)
Comment (pcoet) [reply guides/unified_distribution.py:29]: Or, if "enabling" is too strong, "making it easier for"

File: guides/unified_distribution.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+import keras
+from keras import distribution
+
+"""
+## Introduction
+
+The Keras distribution API is a new API interface designed to facilitate 
+distributed deep learning across a variety of backends like JAX, Tensorflow and
+Pytorch. This innovative API introduces a suite of tools enabling data and model
```
Comment (fchollet) [inline guides/unified_distribution.py:22]: Avoid subjective qualifiers like "innovative". Here maybe "powerful" is better

File: guides/unified_distribution.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+import jax
+import tensorflow as tf     # For dataset input.
+
+"""
+## DeviceMesh and TensorLayout
+
+The DeviceMesh class in Keras distribution API represents a cluster of 
+computational devices configured for distributed computation. It aligns with 
+similar concepts in [jax.sharding.Mesh]
+(https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.Mesh) and
```
Comment (fchollet) [inline guides/unified_distribution.py:68]: Make sure links are unbroken, on a single line (the `](` especially)

File: guides/unified_distribution.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+"""
+## DeviceMesh and TensorLayout
+
+The DeviceMesh class in Keras distribution API represents a cluster of 
+computational devices configured for distributed computation. It aligns with 
+similar concepts in [jax.sharding.Mesh]
+(https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.Mesh) and 
+[tf.dtensor.Mesh]
+(https://www.tensorflow.org/api_docs/python/tf/experimental/dtensor/Mesh),
```
Comment (fchollet) [inline guides/unified_distribution.py:70]: Here too

File: guides/unified_distribution.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+# 
+# The `scope` can also be replaced by `keras.distribution.set_distribution()`, 
+#  which sets the global distribution.
+with data_parallel.scope():
+    inputs = layers.Input(shape=[28, 28, 1])
+    y = layers.Flatten()(inputs)
+    y = layers.Dense(units=200, use_bias=False, activation="relu")(y)
+    y = layers.Dropout(0.4)(y)
+    y = layers.Dense(units=10, activation="softmax")(y)
+    model = models.Model(inputs=inputs, outputs=y)
```
Comment (fchollet) [inline guides/unified_distribution.py:153]: Prefer using `keras.Model`

File: guides/unified_distribution.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+# all the devices of the `DeviceMesh`. This include all the weights like RNG
+# state, optimizer states, metrics, etc. The dataset feed into `model.fit` or
+# `model.evaluate` will be splitted evenly on the batch dimention, and send to
+# all the devices. You don"t have to do any manual aggregration of losses, 
+# since all the computation happens in a global context.
+# 
+# The `scope` can also be replaced by `keras.distribution.set_distribution()`, 
+#  which sets the global distribution.
+with data_parallel.scope():
+    inputs = layers.Input(shape=[28, 28, 1])
```
Comment (fchollet) [inline guides/unified_distribution.py:148]: Use tuples for shapes

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: guides/unified_distribution.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+## Setup
+"""
+
+import keras
+from keras import distribution
+
+"""
+## Introduction
+
+The Keras distribution API is a new API interface designed to facilitate
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+"""
+
+import keras
+from keras import distribution
+
+"""
+## Introduction
+
+The Keras distribution API is a new API interface designed to facilitate 
+distributed deep learning across a variety of backends like JAX, Tensorflow and
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+workflows.
+
+"""
+
+"""
+## How it works
+
+The Keras distribution API provides a global programming model that allows 
+developers to compose applications that operate on Tensors globally while 
+managing the distribution across devices internally. The API leverage the
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+"""
+
+"""
+## How it works
+
+The Keras distribution API provides a global programming model that allows 
+developers to compose applications that operate on Tensors globally while 
+managing the distribution across devices internally. The API leverage the 
+underlying framework to distribute the program and tensors according to the 
+sharding directives through a procedure called Single program, multiple data
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+computational devices configured for distributed computation. It aligns with 
+similar concepts in [jax.sharding.Mesh]
+(https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.Mesh) and 
+[tf.dtensor.Mesh]
+(https://www.tensorflow.org/api_docs/python/tf/experimental/dtensor/Mesh), 
+where it's used to map the physical devices to a logical mesh structure.
+
+The `TensorLayout` class then specifies how tensors are distributed across the
+`DeviceMesh`, detailing the sharding of tensors along specified axes that 
+correspond to the axes names in the `DeviceMesh`.
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+You can found more detailed concepts in [Tensorflow DTensor guide]
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+layout_2d = TensorLayout(axes=("model", "data"), device_mesh=mesh)
+
+# A 4D layout which could be used for data parallel of a image input.
+replicated_layout_4d = TensorLayout(axes=("data", None, None, None), 
+                                    device_mesh=mesh)
+
+"""
+Distribution
+
+The Distribution class in Keras serves as a foundational abstract class designed
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+for developing custom distribution strategies. It encapsulates the core logic 
+needed to distribute a model"s variables, input data, and intermediate
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done. Thanks for catching this. It was changed when I do a find/replace of single quote to double quote.

File: guides/unified_distribution.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+replicated_layout_4d = TensorLayout(axes=("data", None, None, None), 
+                                    device_mesh=mesh)
+
+"""
+Distribution
+
+The Distribution class in Keras serves as a foundational abstract class designed
+for developing custom distribution strategies. It encapsulates the core logic 
+needed to distribute a model"s variables, input data, and intermediate 
+computations across a device mesh. As an end user, you won"t have to interact
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+
+inputs = np.random.normal(size=(128, 28, 28, 1))
+labels = np.random.normal(size=(128, 10))
+dataset = tf.data.Dataset.from_tensor_slices((inputs, labels)).batch(16)
+
+# Note that all the model weights created under the scope are replicated to
+# all the devices of the `DeviceMesh`. This include all the weights like RNG
+# state, optimizer states, metrics, etc. The dataset feed into `model.fit` or
+# `model.evaluate` will be splitted evenly on the batch dimention, and send to
+# all the devices. You don"t have to do any manual aggregration of losses,
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+
+    model.compile(loss="mse")
+    model.fit(dataset, epochs=3)
+    model.evaluate(dataset)
+
+"""
+ModelParallel and LayoutMap
+
+ModelParallel will be mostly useful when model weights are too large to fit
+into a single accelerator. This setting allows you to spit you model weights or
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+with model_parallel.scope():
+    inputs = layers.Input(shape=[28, 28, 1])
+    y = layers.Flatten()(inputs)
+    y = layers.Dense(units=200, use_bias=False, activation="relu", name="d1")(y)
+    y = layers.Dropout(0.4)(y)
+    y = layers.Dense(units=10, activation="softmax", name="d2")(y)
+    model = models.Model(inputs=inputs, outputs=y)
+
+    # The data will be sharded across the "data" dimention of the method, which
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done, and fix for other 3 places as well.

File: guides/unified_distribution.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+# "data" as columns, and it is a [4, 2] grid when it mapped to the physcial
+# devices on the mesh.
+layout_2d = TensorLayout(axes=("model", "data"), device_mesh=mesh)
+
+# A 4D layout which could be used for data parallel of a image input.
+replicated_layout_4d = TensorLayout(axes=("data", None, None, None), 
+                                    device_mesh=mesh)
+
+"""
+Distribution
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+The Distribution class in Keras serves as a foundational abstract class designed
+for developing custom distribution strategies. It encapsulates the core logic 
+needed to distribute a model"s variables, input data, and intermediate 
+computations across a device mesh. As an end user, you won"t have to interact
+directly with this class, but its subclasses like `DataParallel` or 
+`ModelParallel`.
+"""
+
+"""
+DataParallel
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: yes, and fixed.

File: guides/unified_distribution.py [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+    y = layers.Dropout(0.4)(y)
+    y = layers.Dense(units=10, activation="softmax")(y)
+    model = models.Model(inputs=inputs, outputs=y)
+
+    model.compile(loss="mse")
+    model.fit(dataset, epochs=3)
+    model.evaluate(dataset)
+
+"""
+ModelParallel and LayoutMap
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: yes, and fixed.

File: guides/unified_distribution.py [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+making it accessible for developers to accelerate their machine learning 
+workflows.
+
+"""
+
+"""
+## How it works
+
+The Keras distribution API provides a global programming model that allows 
+developers to compose applications that operate on Tensors globally while
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply guides/unified_distribution.py]: done.

File: guides/unified_distribution.py (hunk 1 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py (hunk 2 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py (hunk 3 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py (hunk 4 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done, and fixed for all links.

File: guides/unified_distribution.py (hunk 6 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

File: guides/unified_distribution.py (hunk 7 above)
Comment (qlzh727) [reply guides/unified_distribution.py]: Done.

**Review by pcoet — APPROVED** [review]

**Comment by pcoet** [comment]
@qlzh727 Looks like you made all of the proposed changes. Thanks! I'll go ahead and merge this.

**Comment by qlzh727** [comment]
Great. Thanks.

**Pull request merged**